### PR TITLE
Improve order of evaluation spec, add lifetime of temporaries spec

### PR DESCRIPTION
**Pull request by andralex**
This formalizes behavior already existing in the compiler. It simplifies the rvalue ref DIP.

### CODE REVIEW

**Comment by andralex** [comment]
cc @WalterBright @tgehr @TurkeyMan

File: spec/expression.dd [hunk 1]
```diff
@@ -86,6 +137,85 @@ functions with `extern (D)` $(DDSUBLINK spec/attribute, linkage, linkage).)
     $(BEST_PRACTICE Even though the order of evaluation is well-defined, writing code that
     depends on it is rarely recommended.)
 
+$(H2 $(LNAME2 temporary-lifetime, Lifetime of Temporaries))
+
+$(P Expressions and statements may create and/or consume rvalues. Such values are called
+$(I temporaries) do not have a name or a visible scope. Their lifetime is managed automatically as
```
Comment (dgileadi) [inline spec/expression.dd]: I'm not sure about the grammar in this sentence. Maybe we should insert an "and" after "temporaries"?

Comment (andralex) [reply spec/expression.dd]: Thanks, fixed.

**Comment by thewilsonator** [comment]
`/dev/shm/dtest/work/repo/dlang.org/web/spec/expression.html:144: [RAW MACRO LEAKAGE] $(SPEC_S Expressions,`

**Comment by andralex** [comment]
@thewilsonator found and fixed one missing paren

**Comment by thewilsonator** [comment]
Waiting to see input from Timon. (N.B the 72->h merge does not auto-merge)

File: spec/expression.dd [hunk 2]
```diff
@@ -63,19 +63,70 @@ $(P Note: Rvalues include all literals, special value keywords such as `__FILE__
...
+    writeln("fun() called");
+    static void r(int a, int b, int c) { writeln("callee called"); }
+    return &r;
+}
+int g() { writeln("g() called"); return 1; }
+int h(int x) { writeln("h() called"); return x + 1; }
+int i() { writeln("i() called"); return 2; }
+int j() { writeln("j() called"); return 3; }
+fun()(g(), h(j()), i()); // evaluates fun() then g() then j() then h() then i()
+                         // after which control is transferred to the callee
```
Comment (WalterBright) [inline spec/expression.dd]: swap h and j's names, so you can write the order of evaluation in alphabetical order, which makes it easier to follow.

File: spec/expression.dd [hunk 3]
```diff
@@ -86,6 +137,85 @@ functions with `extern (D)` $(DDSUBLINK spec/attribute, linkage, linkage).)
...
+
+$(P Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+import std.stdio;
+struct S
+{
+    int x;
+    this(int n) { x = n; writefln("~S(%s)", x); }
```
Comment (WalterBright) [inline spec/expression.dd]: Seems to be an extra `~` here in front of `S`

Comment (andralex) [reply spec/expression.dd]: Fixed

File: spec/expression.dd (hunk 2 above)
Comment (andralex) [reply spec/expression.dd]: Fixed

File: spec/expression.dd [hunk 4]
```diff
@@ -63,19 +63,71 @@ $(P Note: Rvalues include all literals, special value keywords such as `__FILE__
...
+void function(int a, int b, int c) fun()
+{
+    writeln("fun() called");
+    static void r(int a, int b, int c) { writeln("callee called"); }
+    return &r;
+}
+int f4() { writeln("f4() called"); return 1; }
+int f3(int x) { writeln("f3() called"); return x + 1; }
+int f1() { writeln("f1() called"); return 2; }
+int f2() { writeln("f2() called"); return 3; }
```
Comment (WalterBright) [inline spec/expression.dd]: Please declare them in order. Also, f4 should return 4, not 1. Etc. By scrambling the order and relationships, it leads the reader to look for something tricky going on, when there isn't.

Comment (andralex) [reply spec/expression.dd]: ah good point

Comment (andralex) [reply spec/expression.dd]: done

Comment (andralex) [reply spec/expression.dd]: oh wait

Comment (andralex) [reply spec/expression.dd]: re-done

**Comment by jacob-carlborg** [comment]
Please make sure the commits are squashed when merging.

**Pull request merged**